lib/pushbox.js

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const Pool = require('./pool')
+const {URLSearchParams} = require('url')
+
+const LOG_OP_RETRIEVE = 'pushbox.retrieve'
+const LOG_OP_STORE = 'pushbox.store'
+
+const SERVICES_TTL = {
+  sendtab: 30 * 24 * 3600 // 30 days.
+}
+
+module.exports = function (log, config) {
+  const pool = new Pool(config.pushboxUrl, { timeout: 15000 })
+
+  return {
+    /**
+     * Retrieves messages for a specific device. Relays the request to Pushbox.
+     *
+     * @param {String} uid - Firefox Account uid
+     * @param {String} deviceId
+     * @param {string} service - Pushbox service identifier
+     * @param {Object} options
+     * @param {String} [options.index]
+     * @param {Number} [options.limit]
+     * @param {String} oAuthToken - Hex string of the user's Pushbox bearer token
+     * @returns {Promise}
+     */
+    retrieve (uid, deviceId, service, {index, limit}, oAuthToken) {
+      log.trace({
+        op: LOG_OP_RETRIEVE,
+        uid,
+        deviceId,
+        service,
+        index,
+        limit
+      })
+      const params = new URLSearchParams()
+      if (index !== undefined) {
+        params.set('index', index)
+      }
+      if (limit !== undefined) {
+        params.set('limit', limit)
+      }
+      return pool.get(`/v1/store/${service}/${uid}/${deviceId}?${params.toString()}`,
+                      { Authorization: `Bearer ${oAuthToken}` })
+    },
+
+    /**
+     * Store service information for a specific device
+     *
+     * @param {String} uid - Firefox Account uid
+     * @param {String} deviceId
+     * @param {string} service - Pushbox service identifier
+     * @param {String} data - Base64 string of the payload
+     * @param {String} oAuthToken - Hex string of the user's Pushbox bearer token
+     * @returns {Promise}
+     */
+    store (uid, deviceId, service, data, oAuthToken) {
+      const ttl = SERVICES_TTL[service]
+      log.trace({
+        op: LOG_OP_STORE,
+        uid,
+        deviceId,
+        service
+      })
+      return pool.post(`/v1/store/${service}/${uid}/${deviceId}`, {
+        data,
+        ttl
+      }, { Authorization: `Bearer ${oAuthToken}` })
```

Comment:
I think it's worth us validating the response here, ensuring we return all and only the fields that we expect to be there.  It might help us detect accidental bustage earlier, and it's a good defense-in-depth measure if either we or pushbox happen to introduce any bugs that could lead to information leakage.

A good approach might be to use a Joi schema, similar to how we validate the claims from the assertion over in oauth-server:

  https://github.com/mozilla/fxa-oauth-server/blob/master/lib/browserid.js#L14

Alternately, we could add response schema validation on the routes and let Hapi do it for us, but IMO it's better to do it as close to the source of the data as possible.